adapters/displayio/displayio.go

```diff
@@ -0,0 +1,207 @@
+package displayio
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/macros"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+	"net/http"
+	"text/template"
+)
+
+type DisplayioAdapter struct {
+	endpoint *template.Template
+}
+
+type reqDioExt struct {
+	UserSession string `json:"userSession,omitempty"`
+	PlacementId string `json:"placementId"`
+	InventoryId string `json:"inventoryId"`
+}
+
+func (adapter *DisplayioAdapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	var requestExt map[string]interface{}
+	var dioExt reqDioExt
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{Message: "No impression in the bid request"}}
+	}
+
+	impressions := request.Imp
+	result := make([]*adapters.RequestData, 0, len(impressions))
+	errs := make([]error, 0, len(impressions))
+
+	for i, impression := range impressions {
+		if impression.BidFloor == 0 {
+			errs = append(errs, &errortypes.BadInput{
+				Message: "BidFloor should be defined",
+			})
+			continue
+		}
+
+		if impression.BidFloorCur == "" {
+			impression.BidFloorCur = "USD"
+		}
+
+		if impression.BidFloorCur != "USD" {
+			convertedValue, err := requestInfo.ConvertCurrency(impression.BidFloor, impression.BidFloorCur, "USD")
+
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			impression.BidFloorCur = "USD"
+			impression.BidFloor = convertedValue
+		}
+
+		if len(impression.Ext) == 0 {
+			errs = append(errs, errors.New("impression extensions required"))
+			continue
+		}
+
+		var bidderExt adapters.ExtImpBidder
+		err := json.Unmarshal(impression.Ext, &bidderExt)
+
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if len(bidderExt.Bidder) == 0 {
+			errs = append(errs, errors.New("bidder required"))
+			continue
+		}
+		var impressionExt openrtb_ext.ExtImpDisplayio
+		err = json.Unmarshal(bidderExt.Bidder, &impressionExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if i == 0 {
+			dioExt = reqDioExt{PlacementId: impressionExt.PlacementId, InventoryId: impressionExt.InventoryId}
+
+			err = json.Unmarshal(request.Ext, &requestExt)
+			if err != nil {
+				requestExt = make(map[string]interface{})
+			}
+
+			requestExt["displayio"] = dioExt
+
+			request.Ext, err = json.Marshal(requestExt)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		}
```

Comment:
Based on adapter code, looks like it is expected to have same placementId and inventoryId for all imps. @xdevel could you confirm if this is expected behaviour. If yes then should specify this in bidder docs.

Reply:
@onkarvhanumante No it's not expected nor required. This code part is rewritten to have separate placement/inventory ids per imp. Thanks.